src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignedCmsTests.netcoreapp.cs

```diff
@@ -225,7 +226,118 @@ public static void SignCmsUsingECDsaCertWithNotMatchingKeyThrows()
+            cms.AddCertificate(newCert);
+
+            Assert.Equal(numOfCerts + 1, cms.Certificates.Count);
+
+            Assert.Equal(1, cms.Certificates.OfType<X509Certificate2>().Where((cert) => cert.Equals(newCert)).Count());
+
+            cms.CheckSignature(true);
+        }
+
+        [Fact]
+        public static void AddCertificateWithPrivateKey()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.CounterSignedRsaPkcs1OneSigner);
+
+            int numOfCerts = cms.Certificates.Count;
+
+            X509Certificate2 newCert = Certificates.RSAKeyTransfer1.TryGetCertificateWithPrivateKey();
+            Assert.True(newCert.HasPrivateKey);
+            cms.AddCertificate(newCert);
+
+            Assert.Equal(numOfCerts + 1, cms.Certificates.Count);
+
+            X509Certificate2 addedCert = cms.Certificates.OfType<X509Certificate2>().Where((cert) => cert.Equals(newCert)).Single();
+            Assert.False(addedCert.HasPrivateKey);
+
+            Assert.Equal(newCert, addedCert);
+
+            cms.CheckSignature(true);
+        }
+
+        [Fact]
+        public static void RemoveCertificate()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.CounterSignedRsaPkcs1OneSigner);
+
+            var expectedCerts = new HashSet<X509Certificate2>(cms.Certificates.OfType<X509Certificate2>());
+
+            X509Certificate2 cert1 = Certificates.RSAKeyTransfer1.GetCertificate();
+            X509Certificate2 cert2 = Certificates.RSAKeyTransfer2.GetCertificate();
+            Assert.NotEqual(cert1, cert2);
+
+            cms.AddCertificate(cert1);
+            cms.AddCertificate(cert2);
+
+            expectedCerts.Add(cert2);
+
+            cms.RemoveCertificate(cert1);
+
+            int count = cms.Certificates.OfType<X509Certificate2>().Where((cert) =>
+                {
+                    Assert.True(expectedCerts.Contains(cert));
+                    return true;
+                }).Count();
+
+            Assert.Equal(expectedCerts.Count, count);
+        }
+
+        [Fact]
+        public static void RemoveNonExistingCertificate()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.CounterSignedRsaPkcs1OneSigner);
+
+            Assert.Throws<CryptographicException>(() => cms.RemoveCertificate(Certificates.RSAKeyTransfer1.GetCertificate()));
+        }
+
+        [Fact]
+        public static void RemoveAllCertsAddBackSignerCert()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.CounterSignedRsaPkcs1OneSigner);
+
+            X509Certificate2 signerCert = cms.SignerInfos[0].Certificate;
+
+            while (cms.Certificates.Count > 0)
+            {
+                cms.RemoveCertificate(cms.Certificates[0]);
+            }
+
+            // Signer info should be gone
+            Assert.Throws<CryptographicException>(() => cms.CheckSignature(true));
```

Comment:
At this point does re-accessing cms.SignerInfos[0].Certificate have the cert, or not?  (A test for behavioral conformance seems good either way).

Prediction:

```C#
Assert.Null(cms.SignerInfos[0].Certificate);
Assert.NotNull(savingTheSignerInfoFromBeforeRemoval.Certificate);
```